contracts/src/mcms.cairo

```diff
@@ -0,0 +1,610 @@
+use starknet::ContractAddress;
+use starknet::{
+    eth_signature::public_key_point_to_eth_address, EthAddress,
+    secp256_trait::{
+        Secp256Trait, Secp256PointTrait, recover_public_key, is_signature_entry_valid, Signature
+    },
+    secp256k1::Secp256k1Point, SyscallResult, SyscallResultTrait
+};
+use alexandria_bytes::{Bytes, BytesTrait};
+use alexandria_encoding::sol_abi::sol_bytes::SolBytesTrait;
+use alexandria_encoding::sol_abi::encode::SolAbiEncodeTrait;
+
+#[starknet::interface]
+trait IManyChainMultiSig<TContractState> {
+    // todo: check length of byte array is 32
+    fn set_root(
+        ref self: TContractState,
+        root: u256,
+        valid_until: u32,
+        metadata: RootMetadata,
+        metadata_proof: Span<u256>,
+        // note: v is a boolean and not uint8
+        signatures: Array<Signature>
+    );
+    fn execute(ref self: TContractState, op: Op, proof: Span<u256>);
+    // // todo: check length of group_quorums and group_parents
+    fn set_config(
+        ref self: TContractState,
+        signer_addresses: Span<EthAddress>,
+        signer_groups: Span<u8>,
+        group_quorums: Span<u8>,
+        group_parents: Span<u8>,
+        clear_root: bool
+    );
+    fn get_config(self: @TContractState) -> Config;
+    fn get_op_count(self: @TContractState) -> u64;
+    fn get_root(self: @TContractState) -> (u256, u32);
+    fn get_root_metadata(self: @TContractState) -> RootMetadata;
+}
+
+#[derive(Copy, Drop, Serde, starknet::Store, PartialEq, Debug)]
+struct Signer {
+    address: EthAddress,
+    index: u8,
+    group: u8
+}
+
+#[derive(Copy, Drop, Serde, starknet::Store, PartialEq)]
+struct RootMetadata {
+    chain_id: u256,
+    multisig: ContractAddress,
+    pre_op_count: u64,
+    post_op_count: u64,
+    override_previous_root: bool
+}
+
+// todo: maybe use copy
+// todo: figure out how this works off-chain with MCMS since we have a new selector field here
```

Comment:
it's doable to make it the first index of the data array, but I figured that the Op struct is already different since it doesn't have "value" in it (nor will sign value offchain) so might as well make it more readable?

Reply:
Have we considered the approach of setting (and asserting) `value == 0`, and thus the Op struct would be exactly the same as EVM? This comes with the benefits of offchain tooling compatibility and consistency across chains, at the cost of readability. 

However I don't think readability (including the additional selector field) is too big of a concern here since the readability of MCM ops comes from proposals describing human-readbale g++ commands, which then get converted into non-human readable raw proposals consisting of smart contract inputs. EVM calldata encodes the function selector in the first 4 bytes of the calldata as well, so its not a foreign concept, and the selector is encoded as a felt meaning its not human-readble in the first place.

That being said, this current approach also works fine as there's no clear winner here - I'll leave it to you to decide on the best approach

Reply:
Yeah I considered that too 🤔 , I prefer readability but also not carying over EVM design decisions to non-EVM chains if they're not absolutely necessary 👍 good thoughts though